### PR TITLE
chore(core): update core to C++17

### PR DESCRIPTION
**Pull request by srl295**
Fixes: #8800

- [x] `std::codecvt` is now deprecated. Switch to ~ICU4C~ _utfcodec.hpp_

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by mcdurdin** [comment]
> * `std::codecvt` is now deprecated. Switch to ~ICU4C~ _utfcodec.hpp_

Good. `std::codecvt` was always weird

**Comment by mcdurdin** [comment]
Cascade of build issues with C++17 changes sadly -- Windows, Developer both failing to build, alongside Core:Windows

**Comment by srl295** [comment]
@mcdurdin doesn't repro locally 😢 

Build server has:

```
   icu-minimal| C++ compiler for the host machine: cl (msvc 19.28.29915 "Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29915 for x86")
18:09:17   icu-minimal| C++ linker for the host machine: link link 14.28.29915.0
```

My system has: `19.29.30146` 

https://learn.microsoft.com/en-us/cpp/overview/compiler-versions?view=msvc-170 says that these numbers indicate VS2019 versions 16.8-16.9 (failing) versus 16.10-16.11 (passing)

**EDIT** Well actually.. repro's locally in x86 but not x64 🤦

**Comment by srl295** [comment]
@mcdurdin summary: works in x64, fails in x86.  The project settings are updated properly.   From long and .. painful experience, sounds like a compiler bug.  Trying VS2022..

**Comment by mcdurdin** [comment]
We could consider eliminating json from C++ code altogether -- as far as I can remember, it's only debug info anyway.

**Comment by srl295** [comment]
> We could consider eliminating json from C++ code altogether -- as far as I can remember, it's only debug info anyway.

it's used for writing debug info and for reading the test data for the ldml runner.

**Comment by srl295** [comment]
I upgraded to VS2022 (only as a test) and it got _worse_ - now x64 breaks too 🤦 

if i upgrade to VS2024 it will make Mac stop working too (kidding)

seriously it seems like it's a known compiler issue on the json side.

**Comment by srl295** [comment]
> I upgraded to VS2022 (only as a test) and it got _worse_ - now x64 breaks too 🤦
> 
> if i upgrade to VS2024 it will make Mac stop working too (kidding)
> 
> seriously it seems like it's a known compiler issue on the json side.

Perhaps a Path Forward is to keep the codecvt stuff out, but build for C++11 for now – maybe have one builder 'try' c++17 on linux to keep it working?

**Comment by mcdurdin** [comment]
> seriously it seems like it's a known compiler issue on the json side.

Any ref to the known issue?

Is there a #define you can use to block certain behaviours of the json lib? I haven't dug in deep but I have seen some examples

**Comment by srl295** [comment]
@mcdurdin wow, the issue was not in `nlohmann::json` (that one was taken care of with the update done earlier), it was in `jsonpp.hpp` which is the non-namespaced `json` class.

it had some fancy templating to handle 'any' kind of string,  but it ended up trying to instantiate a `std::basic_string<const json & ...>` (i.e. inappropriate template expansion).  So rewrite to just be two specializations, one for char16 and one for char.

File: core/src/jsonpp.hpp [hunk 1]
```diff
@@ -113,13 +114,17 @@ json & json::operator << (json::_context_t ctxt) throw()
     return *this;
 }
 
-template<typename C>
 inline
-json & operator << (json & j, std::basic_string<C> const & s) throw() { return j << json::string(convert<C,char>(s).c_str()); }
+json & operator << (json & j, std::string const & s) throw() { return j << json::string(s.c_str()); }
 
-template<typename C>
```
Comment (srl295) [inline core/src/jsonpp.hpp:120]: i think the `C` here matched far too widely.  ANd some of the matches "crashed" with a static assert.

**Review by rc-swag — APPROVED** [review]
lgtm

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.35-alpha](https://keyman.com/downloads/releases/alpha/18.0.35)

**Pull request merged**